pandas/io/parsers.py

```diff
@@ -997,14 +997,14 @@ def _validate_usecols_arg(usecols):
 
     if usecols is not None:
         if callable(usecols):
-            return usecols
+            return usecols, None
         usecols_dtype = lib.infer_dtype(usecols)
         if usecols_dtype not in ('empty', 'integer',
                                  'string', 'unicode'):
             raise ValueError(msg)
 
-        return set(usecols)
-    return usecols
+        return set(usecols), usecols_dtype
```

Comment:
I think it would be more clear to have ``_validate_use_cols`` always return (list, dtype) (for all engines). It makes it very clear what it does. Then have it convert to a ``set``  elsewhere (as you do in ``_set_noconvert_columns``. otherwise you have 2 places which do this conversion (from list -> set), depending  on the engine. again confusing.